// mcporter/dist/daemon/log-context.js
import fsSync from 'node:fs';
import path from 'node:path';
export function createLogContext(options) {
    const derivedEnabled = options.enabled || options.logAllServers || options.servers.size > 0;
    const context = {
        enabled: derivedEnabled,
        logAllServers: options.logAllServers,
        servers: options.servers,
    };
    if (derivedEnabled && options.logPath) {
        try {
            fsSync.mkdirSync(path.dirname(options.logPath), { recursive: true });
            context.writer = fsSync.createWriteStream(options.logPath, {
                flags: 'a',
            });
        }
        catch (error) {
            console.warn(`[daemon] Failed to open log file ${options.logPath}: ${error.message}`);
        }
    }
    return context;
}
export function logEvent(context, message) {
    if (!context.enabled) {
        return;
    }
    const line = `[daemon] ${new Date().toISOString()} ${message}`;
    console.log(line);
    try {
        context.writer?.write(`${line}\n`);
    }
    catch {
        // ignore file write failures
    }
}
export async function disposeLogContext(context) {
    const writer = context.writer;
    if (!writer) {
        return;
    }
    await new Promise((resolve) => {
        writer.end(() => resolve());
        writer.on('error', () => resolve());
    });
}
export function shouldLogServer(context, server) {
    if (!context.enabled) {
        return false;
    }
    if (context.logAllServers) {
        return true;
    }
    return context.servers.has(server);
}
export function formatError(error) {
    if (error instanceof Error) {
        return error.message;
    }
    if (typeof error === 'string') {
        return error;
    }
    return 'unknown';
}
