// mcporter/dist/runtime/transport.js
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
import { StreamableHTTPClientTransport, StreamableHTTPError } from '@modelcontextprotocol/sdk/client/streamableHttp.js';
import { resolveEnvValue, withEnvOverrides } from '../env.js';
import { analyzeConnectionError } from '../error-classifier.js';
import { createOAuthSession } from '../oauth.js';
import { readCachedAccessToken } from '../oauth-persistence.js';
import { materializeHeaders } from '../runtime-header-utils.js';
import { isUnauthorizedError, maybeEnableOAuth } from '../runtime-oauth-support.js';
import { closeTransportAndWait } from '../runtime-process-utils.js';
import { connectWithAuth, isOAuthFlowError, isPostAuthConnectError, OAuthTimeoutError, } from './oauth.js';
import { resolveCommandArgument, resolveCommandArguments } from './utils.js';
const STDIO_TRACE_ENABLED = process.env.MCPORTER_STDIO_TRACE === '1';
function extractTransportStatusCode(error) {
    if (!error || typeof error !== 'object') {
        return undefined;
    }
    const record = error;
    for (const candidate of [record.code, record.status, record.statusCode]) {
        if (typeof candidate === 'number') {
            return candidate;
        }
        if (typeof candidate === 'string') {
            const parsed = Number.parseInt(candidate, 10);
            if (Number.isFinite(parsed)) {
                return parsed;
            }
        }
    }
    return undefined;
}
function isLegacySseTransportMismatch(error) {
    if (error instanceof StreamableHTTPError) {
        return error.code === 404 || error.code === 405;
    }
    const directStatusCode = extractTransportStatusCode(error);
    if (directStatusCode === 404 || directStatusCode === 405) {
        return true;
    }
    const issue = analyzeConnectionError(error);
    return issue.kind === 'http' && (issue.statusCode === 404 || issue.statusCode === 405);
}
function attachStdioTraceLogging(_transport, _label) {
    // STDIO instrumentation is handled via sdk-patches side effects. This helper remains
    // so runtime callers can opt-in without sprinkling conditional checks everywhere.
}
function removeAuthorizationHeader(headers) {
    if (!headers) {
        return undefined;
    }
    for (const key of Object.keys(headers)) {
        if (key.toLowerCase() === 'authorization') {
            delete headers[key];
        }
    }
    return Object.keys(headers).length > 0 ? headers : undefined;
}
function createHttpTransportOptions(definition, oauthSession, shouldEstablishOAuth) {
    const command = definition.command;
    if (command.kind !== 'http') {
        throw new Error(`Server '${definition.name}' is not configured for HTTP transport.`);
    }
    const resolvedHeaders = materializeHeaders(command.headers, definition.name);
    const effectiveHeaders = shouldEstablishOAuth ? removeAuthorizationHeader(resolvedHeaders) : resolvedHeaders;
    return {
        requestInit: effectiveHeaders ? { headers: effectiveHeaders } : undefined,
        authProvider: oauthSession?.provider,
    };
}
async function closeOAuthSession(oauthSession) {
    await oauthSession?.close().catch(() => { });
}
function shouldAbortSseFallback(error) {
    if (isPostAuthConnectError(error)) {
        return !isLegacySseTransportMismatch(error);
    }
    return isOAuthFlowError(error) || error instanceof OAuthTimeoutError;
}
function maybePromoteHttpDefinition(definition, logger, options) {
    if (options.maxOAuthAttempts === 0) {
        return undefined;
    }
    return maybeEnableOAuth(definition, logger);
}
async function connectHttpTransport(client, transport, oauthSession, logger, connectOptions) {
    try {
        return (await connectWithAuth(client, transport, oauthSession, logger, connectOptions));
    }
    catch (error) {
        await closeTransportAndWait(logger, transport).catch(() => { });
        throw error;
    }
}
async function applyCachedOAuthHeaderIfAvailable(definition, logger, allowCachedAuth) {
    if (!allowCachedAuth || definition.auth !== 'oauth' || definition.command.kind !== 'http') {
        return definition;
    }
    try {
        const cached = await readCachedAccessToken(definition, logger);
        if (!cached) {
            return definition;
        }
        const existingHeaders = definition.command.headers ?? {};
        if ('Authorization' in existingHeaders) {
            return definition;
        }
        logger.debug?.(`Using cached OAuth access token for '${definition.name}' (non-interactive).`);
        return {
            ...definition,
            command: {
                ...definition.command,
                headers: {
                    ...existingHeaders,
                    Authorization: `Bearer ${cached}`,
                },
            },
        };
    }
    catch (error) {
        logger.debug?.(`Failed to read cached OAuth token for '${definition.name}': ${error instanceof Error ? error.message : String(error)}`);
        return definition;
    }
}
async function createStdioClientContext(client, definition, logger) {
    const resolvedEnvOverrides = definition.env && Object.keys(definition.env).length > 0
        ? Object.fromEntries(Object.entries(definition.env)
            .map(([key, raw]) => [key, resolveEnvValue(raw)])
            .filter(([, value]) => value !== ''))
        : undefined;
    const mergedEnv = resolvedEnvOverrides && Object.keys(resolvedEnvOverrides).length > 0
        ? { ...process.env, ...resolvedEnvOverrides }
        : { ...process.env };
    const transport = new StdioClientTransport({
        command: resolveCommandArgument(definition.command.command),
        args: resolveCommandArguments(definition.command.args),
        cwd: definition.command.cwd,
        env: mergedEnv,
    });
    if (STDIO_TRACE_ENABLED) {
        attachStdioTraceLogging(transport, definition.name ?? definition.command.command);
    }
    try {
        await client.connect(transport);
    }
    catch (error) {
        await closeTransportAndWait(logger, transport).catch(() => { });
        throw error;
    }
    return { client, transport, definition, oauthSession: undefined };
}
async function retryHttpTransportWithFallback(client, definition, logger, options) {
    let activeDefinition = definition;
    while (true) {
        const attempt = await attemptHttpClientContext(client, activeDefinition, logger, options);
        if (!attempt.nextDefinition) {
            return attempt.context;
        }
        activeDefinition = attempt.nextDefinition;
        options.onDefinitionPromoted?.(activeDefinition);
    }
}
async function attemptHttpClientContext(client, activeDefinition, logger, options) {
    const command = activeDefinition.command;
    if (command.kind !== 'http') {
        throw new Error(`Server '${activeDefinition.name}' is not configured for HTTP transport.`);
    }
    let oauthSession;
    const shouldEstablishOAuth = activeDefinition.auth === 'oauth' && options.maxOAuthAttempts !== 0;
    if (shouldEstablishOAuth) {
        oauthSession = await createOAuthSession(activeDefinition, logger);
    }
    const transportOptions = createHttpTransportOptions(activeDefinition, oauthSession, shouldEstablishOAuth);
    try {
        const context = await connectPrimaryHttpTransport(client, activeDefinition, command, transportOptions, oauthSession, logger, options);
        return { context };
    }
    catch (primaryError) {
        if (shouldAbortSseFallback(primaryError)) {
            await closeOAuthSession(oauthSession);
            throw primaryError;
        }
        if (isUnauthorizedError(primaryError)) {
            await closeOAuthSession(oauthSession);
            const promoted = maybePromoteHttpDefinition(activeDefinition, logger, options);
            if (promoted) {
                return { nextDefinition: promoted };
            }
            oauthSession = undefined;
        }
        if (primaryError instanceof Error) {
            logger.info(`Falling back to SSE transport for '${activeDefinition.name}': ${primaryError.message}`);
        }
        return {
            context: await connectSseFallbackTransport(client, activeDefinition, command, transportOptions, oauthSession, logger, options),
        };
    }
}
async function connectPrimaryHttpTransport(client, definition, command, transportOptions, oauthSession, logger, options) {
    const createStreamableTransport = () => new StreamableHTTPClientTransport(command.url, transportOptions);
    const transport = await connectHttpTransport(client, createStreamableTransport(), oauthSession, logger, {
        serverName: definition.name,
        maxAttempts: options.maxOAuthAttempts,
        oauthTimeoutMs: options.oauthTimeoutMs,
        recreateTransport: async () => createStreamableTransport(),
    });
    return {
        client,
        transport,
        definition,
        oauthSession,
    };
}
async function connectSseFallbackTransport(client, definition, command, transportOptions, oauthSession, logger, options) {
    try {
        const transport = await connectHttpTransport(client, new SSEClientTransport(command.url, transportOptions), oauthSession, logger, {
            serverName: definition.name,
            maxAttempts: options.maxOAuthAttempts,
            oauthTimeoutMs: options.oauthTimeoutMs,
        });
        return { client, transport, definition, oauthSession };
    }
    catch (sseError) {
        await closeOAuthSession(oauthSession);
        if (sseError instanceof OAuthTimeoutError) {
            throw sseError;
        }
        if (isUnauthorizedError(sseError)) {
            const promoted = maybePromoteHttpDefinition(definition, logger, options);
            if (promoted) {
                options.onDefinitionPromoted?.(promoted);
                return retryHttpTransportWithFallback(client, promoted, logger, options);
            }
        }
        throw sseError;
    }
}
export async function createClientContext(definition, logger, clientInfo, options = {}) {
    const client = new Client(clientInfo);
    const activeDefinition = await applyCachedOAuthHeaderIfAvailable(definition, logger, options.allowCachedAuth);
    return withEnvOverrides(activeDefinition.env, async () => {
        if (activeDefinition.command.kind === 'stdio') {
            return createStdioClientContext(client, activeDefinition, logger);
        }
        return retryHttpTransportWithFallback(client, activeDefinition, logger, options);
    });
}
