// mcporter/dist/cli/call-arguments.js
import { parseLeadingCallExpression } from './call-argument-expression.js';
import { coerceValue, parseKeyValueToken, shouldPromoteSelectorToCommand, } from './call-argument-values.js';
import { buildUnknownCallFlagMessage } from './call-help.js';
import { extractEphemeralServerFlags } from './ephemeral-flags.js';
import { CliUsageError } from './errors.js';
import { consumeOutputFormat } from './output-format.js';
import { consumeTimeoutFlag } from './timeouts.js';
const FLAG_HANDLERS = {
    '--server': handleServerFlag,
    '--mcp': handleServerFlag,
    '--tool': handleToolFlag,
    '--timeout': handleTimeoutFlag,
    '--tail-log': handleTailLogFlag,
    '--save-images': handleSaveImagesFlag,
    '--yes': handleNoopFlag,
    '--raw-strings': handleRawStringsFlag,
    '--no-coerce': handleNoCoerceFlag,
    '--args': handleArgsFlag,
};
export function parseCallArguments(args) {
    const result = { args: {}, tailLog: false, output: 'auto' };
    const flagState = { coercionMode: 'default' };
    const ephemeral = extractEphemeralServerFlags(args);
    result.ephemeral = ephemeral;
    result.output = consumeOutputFormat(args, {
        defaultFormat: 'auto',
    });
    const { positional, literalPositional } = scanCallTokens(args, result, flagState);
    const { callExpressionProvidedServer, callExpressionProvidedTool } = applyLeadingCallExpression(positional, result);
    resolveSelectorAndTool(positional, result, callExpressionProvidedServer, callExpressionProvidedTool);
    applyTrailingArguments(positional, result, flagState);
    appendLiteralPositionalArguments(literalPositional, result, flagState);
    return result;
}
function scanCallTokens(args, result, state) {
    const positional = [];
    const literalPositional = [];
    let index = 0;
    while (index < args.length) {
        const token = args[index];
        if (!token) {
            index += 1;
            continue;
        }
        if (token === '--') {
            literalPositional.push(...args.slice(index + 1).filter(Boolean));
            break;
        }
        const flagHandler = FLAG_HANDLERS[token];
        if (flagHandler) {
            index = flagHandler({ args, index, result, state });
            continue;
        }
        if (token.startsWith('--')) {
            throw new CliUsageError(buildUnknownCallFlagMessage(token));
        }
        positional.push(token);
        index += 1;
    }
    return { positional, literalPositional };
}
function applyLeadingCallExpression(positional, result) {
    if (positional.length === 0) {
        return { callExpressionProvidedServer: false, callExpressionProvidedTool: false };
    }
    const rawToken = positional[0] ?? '';
    const callExpression = parseLeadingCallExpression(rawToken);
    if (!callExpression) {
        return { callExpressionProvidedServer: false, callExpressionProvidedTool: false };
    }
    positional.shift();
    if (callExpression.server) {
        if (result.server && result.server !== callExpression.server) {
            throw new Error(`Conflicting server names: '${result.server}' from flags and '${callExpression.server}' from call expression.`);
        }
        result.server = result.server ?? callExpression.server;
    }
    if (result.tool && result.tool !== callExpression.tool) {
        throw new Error(`Conflicting tool names: '${result.tool}' from flags and '${callExpression.tool}' from call expression.`);
    }
    result.tool = callExpression.tool;
    Object.assign(result.args, callExpression.args);
    if (callExpression.positionalArgs && callExpression.positionalArgs.length > 0) {
        result.positionalArgs = [...(result.positionalArgs ?? []), ...callExpression.positionalArgs];
    }
    return {
        callExpressionProvidedServer: Boolean(callExpression.server),
        callExpressionProvidedTool: Boolean(callExpression.tool),
    };
}
function resolveSelectorAndTool(positional, result, callExpressionProvidedServer, callExpressionProvidedTool) {
    if (!result.selector && positional.length > 0 && !callExpressionProvidedServer && !result.server) {
        result.selector = positional.shift();
    }
    if (!result.server &&
        result.selector &&
        shouldPromoteSelectorToCommand(result.selector) &&
        !result.ephemeral?.stdioCommand) {
        result.ephemeral = { ...result.ephemeral, stdioCommand: result.selector };
        result.selector = undefined;
    }
    const nextPositional = positional[0];
    if (!result.tool &&
        nextPositional !== undefined &&
        !nextPositional.includes('=') &&
        !nextPositional.includes(':') &&
        !callExpressionProvidedTool) {
        result.tool = positional.shift();
    }
}
function applyTrailingArguments(positional, result, state) {
    const trailingPositional = [];
    for (let index = 0; index < positional.length;) {
        const token = positional[index];
        if (!token) {
            index += 1;
            continue;
        }
        const parsed = parseKeyValueToken(token, positional[index + 1]);
        if (!parsed) {
            trailingPositional.push(coerceValue(token, state.coercionMode));
            index += 1;
            continue;
        }
        index += parsed.consumed;
        const value = coerceValue(parsed.rawValue, state.coercionMode);
        if (parsed.key === 'tool' && !result.tool) {
            if (typeof value !== 'string') {
                throw new Error("Argument 'tool' must be a string value.");
            }
            result.tool = value;
            continue;
        }
        if (parsed.key === 'server' && !result.server) {
            if (typeof value !== 'string') {
                throw new Error("Argument 'server' must be a string value.");
            }
            result.server = value;
            continue;
        }
        result.args[parsed.key] = value;
    }
    if (trailingPositional.length > 0) {
        result.positionalArgs = [...(result.positionalArgs ?? []), ...trailingPositional];
    }
}
function appendLiteralPositionalArguments(literalPositional, result, state) {
    if (literalPositional.length === 0) {
        return;
    }
    result.positionalArgs = [
        ...(result.positionalArgs ?? []),
        ...literalPositional.map((token) => coerceValue(token, state.coercionMode)),
    ];
}
function handleServerFlag(context) {
    const token = context.args[context.index] ?? '--server';
    context.result.server = consumeFlagValue(context.args, context.index, token);
    return context.index + 2;
}
function handleToolFlag(context) {
    context.result.tool = consumeFlagValue(context.args, context.index, '--tool');
    return context.index + 2;
}
function handleTimeoutFlag(context) {
    context.result.timeoutMs = consumeTimeoutFlag(context.args, context.index, {
        flagName: '--timeout',
        missingValueMessage: '--timeout requires a value (milliseconds).',
    });
    // consumeTimeoutFlag removes the flag/value pair in-place; stay on the same index.
    return context.index;
}
function handleTailLogFlag(context) {
    context.result.tailLog = true;
    return context.index + 1;
}
function handleSaveImagesFlag(context) {
    context.result.saveImagesDir = consumeFlagValue(context.args, context.index, '--save-images', '--save-images requires a directory path.');
    return context.index + 2;
}
function handleNoopFlag(context) {
    return context.index + 1;
}
function handleRawStringsFlag(context) {
    context.state.coercionMode = 'raw-strings';
    context.result.rawStrings = true;
    return context.index + 1;
}
function handleNoCoerceFlag(context) {
    context.state.coercionMode = 'none';
    context.result.rawStrings = true;
    return context.index + 1;
}
function handleArgsFlag(context) {
    const raw = consumeFlagValue(context.args, context.index, '--args', '--args requires a JSON value.');
    let decoded;
    try {
        decoded = JSON.parse(raw);
    }
    catch (error) {
        throw new Error(`Unable to parse --args: ${error.message}`);
    }
    if (decoded === null || typeof decoded !== 'object' || Array.isArray(decoded)) {
        throw new Error('Unable to parse --args: --args must be a JSON object.');
    }
    Object.assign(context.result.args, decoded);
    return context.index + 2;
}
function consumeFlagValue(args, index, token, missingValueMessage) {
    const value = args[index + 1];
    if (value) {
        return value;
    }
    throw new Error(missingValueMessage ?? `Flag '${token}' requires a value.`);
}
