// mcporter/dist/cli.js
#!/usr/bin/env node
import { handleAuth, printAuthHelp } from './cli/auth-command.js';
import { printCallHelp, handleCall as runHandleCall } from './cli/call-command.js';
import { buildGlobalContext } from './cli/cli-factory.js';
import { inferCommandRouting } from './cli/command-inference.js';
import { handleConfigCli } from './cli/config-command.js';
import { handleDaemonCli } from './cli/daemon-command.js';
import { handleEmitTs } from './cli/emit-ts-command.js';
import { CliUsageError } from './cli/errors.js';
import { handleGenerateCli } from './cli/generate-cli-runner.js';
import { consumeHelpTokens, isHelpToken, isVersionToken, printHelp, printVersion } from './cli/help-output.js';
import { handleInspectCli } from './cli/inspect-cli-command.js';
import { handleList, printListHelp } from './cli/list-command.js';
import { logError, logInfo } from './cli/logger-context.js';
import { DEBUG_HANG, dumpActiveHandles, terminateChildProcesses } from './cli/runtime-debug.js';
import { resolveConfigPath } from './config.js';
import { DaemonClient } from './daemon/client.js';
import { createKeepAliveRuntime } from './daemon/runtime-wrapper.js';
import { isKeepAliveServer } from './lifecycle.js';
import { createRuntime } from './runtime.js';
export { handleAuth, printAuthHelp } from './cli/auth-command.js';
export { parseCallArguments } from './cli/call-arguments.js';
export { handleCall } from './cli/call-command.js';
export { handleGenerateCli } from './cli/generate-cli-runner.js';
export { handleInspectCli } from './cli/inspect-cli-command.js';
export { extractListFlags, handleList } from './cli/list-command.js';
export { resolveCallTimeout } from './cli/timeouts.js';
export async function runCli(argv) {
    const args = [...argv];
    if (args.length === 0) {
        printHelp();
        process.exit(1);
        return;
    }
    const context = buildGlobalContext(args);
    if ('exit' in context) {
        process.exit(context.code);
        return;
    }
    const { globalFlags, runtimeOptions } = context;
    const command = args.shift();
    if (!command) {
        printHelp();
        process.exit(1);
        return;
    }
    if (isHelpToken(command)) {
        printHelp();
        process.exitCode = 0;
        return;
    }
    if (isVersionToken(command)) {
        await printVersion();
        return;
    }
    // Early-exit command handlers that don't require runtime inference.
    if (command === 'generate-cli') {
        await handleGenerateCli(args, globalFlags);
        return;
    }
    if (command === 'inspect-cli') {
        await handleInspectCli(args);
        return;
    }
    const rootOverride = globalFlags['--root'];
    const configPath = runtimeOptions.configPath ?? globalFlags['--config'];
    const configResolution = resolveConfigPath(globalFlags['--config'], rootOverride ?? process.cwd());
    const configPathResolved = configPath ?? configResolution.path;
    // Only pass configPath to runtime options if it was explicitly provided (via --config flag or env var).
    // If not explicit, let loadConfigLayers handle the default resolution to avoid ENOENT on missing config.
    const runtimeOptionsWithPath = {
        ...runtimeOptions,
        configPath: configResolution.explicit ? configPathResolved : runtimeOptions.configPath,
    };
    if (command === 'daemon') {
        await handleDaemonCli(args, {
            configPath: configPathResolved,
            configExplicit: configResolution.explicit,
            rootDir: rootOverride,
        });
        return;
    }
    if (command === 'config') {
        await handleConfigCli({
            loadOptions: { configPath, rootDir: rootOverride },
            invokeAuth: (authArgs) => invokeAuthCommand(runtimeOptionsWithPath, authArgs),
        }, args);
        return;
    }
    if (command === 'emit-ts') {
        const runtime = await createRuntime(runtimeOptionsWithPath);
        try {
            await handleEmitTs(runtime, args);
        }
        finally {
            await runtime.close().catch(() => { });
        }
        return;
    }
    const baseRuntime = await createRuntime(runtimeOptionsWithPath);
    const keepAliveServers = new Set(baseRuntime
        .getDefinitions()
        .filter(isKeepAliveServer)
        .map((entry) => entry.name));
    const daemonClient = keepAliveServers.size > 0
        ? new DaemonClient({
            configPath: configResolution.path,
            configExplicit: configResolution.explicit,
            rootDir: rootOverride,
        })
        : null;
    const runtime = createKeepAliveRuntime(baseRuntime, { daemonClient, keepAliveServers });
    const inference = inferCommandRouting(command, args, runtime.getDefinitions());
    if (inference.kind === 'abort') {
        process.exitCode = inference.exitCode;
        return;
    }
    const resolvedCommand = inference.command;
    const resolvedArgs = inference.args;
    try {
        if (resolvedCommand === 'list') {
            if (consumeHelpTokens(resolvedArgs)) {
                printListHelp();
                process.exitCode = 0;
                return;
            }
            await handleList(runtime, resolvedArgs);
            return;
        }
        if (resolvedCommand === 'call') {
            if (consumeHelpTokens(resolvedArgs)) {
                printCallHelp();
                process.exitCode = 0;
                return;
            }
            await runHandleCall(runtime, resolvedArgs);
            return;
        }
        if (resolvedCommand === 'auth') {
            if (consumeHelpTokens(resolvedArgs)) {
                printAuthHelp();
                process.exitCode = 0;
                return;
            }
            await handleAuth(runtime, resolvedArgs);
            return;
        }
    }
    finally {
        const closeStart = Date.now();
        if (DEBUG_HANG) {
            logInfo('[debug] beginning runtime.close()');
            dumpActiveHandles('before runtime.close');
        }
        try {
            await runtime.close();
            if (DEBUG_HANG) {
                const duration = Date.now() - closeStart;
                logInfo(`[debug] runtime.close() completed in ${duration}ms`);
                dumpActiveHandles('after runtime.close');
            }
        }
        catch (error) {
            if (DEBUG_HANG) {
                logError('[debug] runtime.close() failed', error);
            }
        }
        finally {
            terminateChildProcesses('runtime.finally');
            // By default we force an exit after cleanup so Node doesn't hang on lingering stdio handles
            // (see typescript-sdk#579/#780/#1049). Opt out by exporting MCPORTER_NO_FORCE_EXIT=1.
            const disableForceExit = process.env.MCPORTER_NO_FORCE_EXIT === '1';
            if (DEBUG_HANG) {
                dumpActiveHandles('after terminateChildProcesses');
                if (!disableForceExit || process.env.MCPORTER_FORCE_EXIT === '1') {
                    process.exit(0);
                }
            }
            else {
                const scheduleExit = () => {
                    if (!disableForceExit || process.env.MCPORTER_FORCE_EXIT === '1') {
                        process.exit(0);
                    }
                };
                setImmediate(scheduleExit);
            }
        }
    }
    printHelp(`Unknown command '${resolvedCommand}'.`);
    process.exit(1);
}
// main parses CLI flags and dispatches to list/call commands.
async function main() {
    await runCli(process.argv.slice(2));
}
if (process.env.MCPORTER_DISABLE_AUTORUN !== '1') {
    main().catch((error) => {
        if (error instanceof CliUsageError) {
            logError(error.message);
            process.exit(1);
            return;
        }
        const message = error instanceof Error ? error.message : String(error);
        logError(message, error instanceof Error ? error : undefined);
        process.exit(1);
    });
}
async function invokeAuthCommand(runtimeOptions, args) {
    const runtime = await createRuntime(runtimeOptions);
    try {
        await handleAuth(runtime, args);
    }
    finally {
        await runtime.close().catch(() => { });
    }
}
