// mcporter/dist/result-utils.js
import { analyzeConnectionError } from './error-classifier.js';
function extractEnvelope(raw) {
    if (!raw || typeof raw !== 'object') {
        return { content: null, structuredContent: null };
    }
    const obj = raw;
    let content = null;
    let structuredContent = null;
    if ('content' in obj && Array.isArray(obj.content)) {
        content = obj.content;
    }
    if ('structuredContent' in obj) {
        structuredContent = obj.structuredContent;
    }
    if ('raw' in obj && obj.raw && typeof obj.raw === 'object') {
        const nested = obj.raw;
        if (!content && 'content' in nested && Array.isArray(nested.content)) {
            content = nested.content;
        }
        if (structuredContent === null && 'structuredContent' in nested) {
            structuredContent = nested.structuredContent;
        }
    }
    return { content, structuredContent };
}
// asString converts known content/value shapes into plain strings.
function asString(value) {
    if (typeof value === 'string') {
        return value;
    }
    if (value && typeof value === 'object' && 'text' in value) {
        const text = value.text;
        return typeof text === 'string' ? text : null;
    }
    return null;
}
function collectCallContent(raw) {
    const envelope = extractEnvelope(raw);
    const textEntries = [];
    const markdownEntries = [];
    const jsonCandidates = [];
    const images = [];
    if (!envelope.content) {
        return {
            content: envelope.content,
            structuredContent: envelope.structuredContent,
            textEntries,
            markdownEntries,
            jsonCandidates,
            images,
        };
    }
    for (const entry of envelope.content) {
        if (typeof entry === 'string') {
            const parsed = tryParseJson(entry);
            if (parsed !== null) {
                jsonCandidates.push(parsed);
            }
            continue;
        }
        if (!entry || typeof entry !== 'object' || !('type' in entry)) {
            continue;
        }
        const typedEntry = entry;
        if (typedEntry.type === 'json') {
            const parsed = tryParseJson(entry);
            if (parsed !== null) {
                jsonCandidates.push(parsed);
            }
            continue;
        }
        if (typedEntry.type === 'image') {
            const data = typedEntry.data;
            const mimeType = typedEntry.mimeType ?? 'image/png';
            if (typeof data === 'string' && typeof mimeType === 'string') {
                images.push({ data, mimeType });
            }
            continue;
        }
        if (typedEntry.type === 'resource') {
            const resource = typedEntry.resource;
            if (resource && typeof resource === 'object') {
                const uri = typeof resource.uri === 'string' ? resource.uri : '';
                const mimeType = typeof resource.mimeType === 'string' ? resource.mimeType : '';
                if (typeof resource.text === 'string') {
                    textEntries.push(resource.text);
                    if (mimeType.toLowerCase().includes('markdown')) {
                        markdownEntries.push(resource.text);
                    }
                    const parsed = tryParseJson(resource.text);
                    if (parsed !== null) {
                        jsonCandidates.push(parsed);
                    }
                }
                else if (typeof resource.blob === 'string') {
                    textEntries.push(`[Binary resource: ${uri}]`);
                }
            }
            continue;
        }
        if (typedEntry.type !== 'text' && typedEntry.type !== 'markdown') {
            continue;
        }
        const text = asString(entry);
        if (!text) {
            continue;
        }
        textEntries.push(text);
        if (typedEntry.type === 'markdown') {
            markdownEntries.push(text);
        }
        const parsed = tryParseJson(text);
        if (parsed !== null) {
            jsonCandidates.push(parsed);
        }
    }
    return {
        content: envelope.content,
        structuredContent: envelope.structuredContent,
        textEntries,
        markdownEntries,
        jsonCandidates,
        images,
    };
}
function collectText(entries, joiner) {
    if (entries.length === 0) {
        return null;
    }
    return entries.join(joiner);
}
function collectImages(images) {
    if (images.length === 0) {
        return null;
    }
    return images;
}
function unwrapJsonEnvelope(record, fallback) {
    if ('json' in record) {
        return record.json ?? null;
    }
    if ('data' in record) {
        return Object.keys(record).length === 1 ? (record.data ?? null) : fallback;
    }
    return null;
}
function parseStructuredContent(value) {
    if (value === undefined || value === null) {
        return null;
    }
    if (typeof value === 'string') {
        return tryParseJson(value);
    }
    if (typeof value !== 'object') {
        return null;
    }
    const unwrapped = unwrapJsonEnvelope(value, value);
    return unwrapped ?? value;
}
// tryParseJson pulls JSON payloads out of structured responses or raw strings.
function tryParseJson(value) {
    if (value === undefined || value === null) {
        return null;
    }
    if (typeof value === 'object') {
        const unwrapped = unwrapJsonEnvelope(value, value);
        if (unwrapped !== null) {
            return unwrapped;
        }
    }
    if (typeof value === 'string') {
        try {
            return JSON.parse(value);
        }
        catch {
            return null;
        }
    }
    return null;
}
// createCallResult wraps a tool response with helpers for common content types.
export function createCallResult(raw) {
    let cachedContent;
    const getCollectedContent = () => {
        if (cachedContent) {
            return cachedContent;
        }
        cachedContent = collectCallContent(raw);
        return cachedContent;
    };
    return {
        raw,
        text(joiner = '\n') {
            if (raw == null) {
                return null;
            }
            if (typeof raw === 'string') {
                return raw;
            }
            const collected = getCollectedContent();
            const combinedText = collectText(collected.textEntries, joiner);
            if (combinedText) {
                return combinedText;
            }
            return asString(collected.structuredContent);
        },
        markdown(joiner = '\n') {
            const collected = getCollectedContent();
            const structured = collected.structuredContent;
            if (structured && typeof structured === 'object') {
                const markdown = structured.markdown;
                if (typeof markdown === 'string') {
                    return markdown;
                }
            }
            return collectText(collected.markdownEntries, joiner);
        },
        json() {
            const collected = getCollectedContent();
            const parsedStructured = parseStructuredContent(collected.structuredContent);
            if (parsedStructured !== null) {
                return parsedStructured;
            }
            if (collected.jsonCandidates.length === 1) {
                return collected.jsonCandidates[0];
            }
            if (collected.jsonCandidates.length > 1) {
                return collected.jsonCandidates;
            }
            if (typeof raw === 'string') {
                const parsedRaw = tryParseJson(raw);
                if (parsedRaw !== null) {
                    return parsedRaw;
                }
            }
            const textContent = this.text?.();
            if (typeof textContent === 'string') {
                const parsedText = tryParseJson(textContent);
                if (parsedText !== null) {
                    return parsedText;
                }
            }
            const markdownContent = this.markdown?.();
            if (typeof markdownContent === 'string') {
                const parsedMarkdown = tryParseJson(markdownContent);
                if (parsedMarkdown !== null) {
                    return parsedMarkdown;
                }
            }
            return null;
        },
        images() {
            const collected = getCollectedContent();
            return collectImages(collected.images);
        },
        content() {
            return getCollectedContent().content;
        },
        structuredContent() {
            return getCollectedContent().structuredContent;
        },
    };
}
export function describeConnectionIssue(error) {
    return analyzeConnectionError(error);
}
export function wrapCallResult(raw) {
    return { raw, callResult: createCallResult(raw) };
}
